Tuples interfaces generics (#52891)
dotnet/docs: docs/csharp/fundamentals/types/generics.md

You encounter generics constantly in everyday C#. Collections, async return types, delegates, and LINQ all rely on generic types:

:::code language="csharp" source="snippets/generics/Program.cs" ID="EverydayGenerics":::

In each case, the type argument in angle brackets (`<int>`, `<string>`, `<Product>`) tells the generic type what kind of data it holds or operates on. The compiler enforces type safety—you can't accidentally add a `string` to a `List<int>`.

## Consuming generic types

Most of the time, you *consume* generic types from the .NET class library rather than creating your own. The following sections show the most common generic types you'll use.

### Generic collections

The <xref:System.Collections.Generic> namespace provides type-safe collection classes. Always use these instead of nongeneric collections like <xref:System.Collections.ArrayList>:

:::code language="csharp" source="snippets/generics/Program.cs" ID="GenericCollections":::

> Generic collections prevent type errors at compile time rather than at runtime. They also avoid boxing for value types, which improves performance.

Reviewer: This doesn't read quite right as is.

```suggestion
Generic collections prevent type errors at runtime because the errors surface at compile time instead. These collections also avoid boxing for value types, which improves performance.
```
